src/pci_config.rs

```
@@ -0,0 +1,714 @@
+pub const SUBCLASS_OFFSET: usize = 0x0A;
+/// Offset of class code in the PCI configuration space.
+pub const CLASS_CODE_OFFSET: usize = 0x0B;
+/// Offset of cache line size in the PCI configuration space.
+pub const CACHE_LINE_SIZE_OFFSET: usize = 0x0C;
+/// Offset of latency timer in the PCI configuration space.
+pub const LATENCY_TIMER_OFFSET: usize = 0x0D;
+/// Offset of header type in the PCI configuration space.
+pub const HEADER_TYPE_OFFSET: usize = 0x0E;
+/// Offset of BIST in the PCI configuration space.
+pub const BIST_OFFSET: usize = 0x0F;
+
+// Check to ensure all address accesses to the configuration space are 32bit
+// aligned.
+pub(crate) fn validate_dword_alignment(addr: usize) -> Result<()> {
+    if addr % 4 != 0 {
+        return Err(Error::UnalignedAccess);
+    }
+    Ok(())
+}
+
+#[derive(Debug)]
+/// Error type for PCI configuration space accesses.
+pub enum Error {
+    /// Offset of the access to the configuration space has improper alignment.
+    UnalignedAccess,
+    /// Offset of the access to the configuration space is out of bounds.
+    OffsetOutOfBounds,
+    // Length of the buffer in the access to the configuration space is
+    // invalid.
+    // Only used if `read` and `write` methods use buffers and not `u32`s.
+
+    // InvalidDataLen,
+}
+
+impl std::error::Error for Error {}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use Error::*;
+
+        match *self {
+            UnalignedAccess => write!(f, "unaligned access"),
+            OffsetOutOfBounds => write!(f, "out of bounds access"),
+        }
+    }
+}
+
+/// Enum representing PCI header types as presented in the PCI specification.
+#[derive(Copy, Clone)]
+pub enum PciHeaderType {
+    /// Generic PCI device.
+    Device = 0x00,
+    /// PCI-to-PCI bridge.
+    PciToPciBridge = 0x01,
+    /// PCI-to-Cardbus bridge.
+    PciToCardbusBridge = 0x02,
+    /// Unknown header type.
```

According to the spec, all the other values beside 0, 1, 2 are reserved, and so there are not going to be defined in valid PCI headers that are implementing the 3.0 spec. Thus, it would be better to not explicitly define this here so that error cases can be explicit in the code that parses the header. Otherwise, users need to validate that the header type is not "Unknown". To make this possible, the below implementation of `From<u8> for PciHeaderType` should be converted to a `TryFrom<u8>` such that we can return an error in case the Header type is invalid.

---

Same should happen for `PciClassCode`.